### PR TITLE
Widen `RefKindVector` to hold `RefReadonlyParameter`

### PR DESCRIPTION
**Pull request by jjonescz**
Follow up on https://github.com/dotnet/roslyn/pull/68179#discussion_r1224607989.
Test plan: https://github.com/dotnet/roslyn/issues/68056

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LoweredDynamicOperationFactory.cs [hunk 1]
```diff
@@ -767,7 +767,6 @@ internal FieldSymbol DefineCallSiteStorageSymbol(NamedTypeSymbol containerDefini
                 }
             }
 
-            // PROTOTYPE: RefKindVector does not support RefReadOnlyParameter.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LoweredDynamicOperationFactory.cs:770]: >// PROTOTYPE: RefKindVector does not support RefReadOnlyParameter. [](http://example.com/codeflow?start=12&length=66)

Are we testing this code path when `ref readonly` parameter is involved?

**Comment by AlekseyTs** [comment]
>         public static bool TryParse(string refKindString, int capacity, out RefKindVector result)

Do we need to worry about change in interpretation across compiler versions? Is there a scenario when an old compiler has to parse a name generated by new compiler and vise versa?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Synthesized/RefKindVector.cs:109 in 79e0e84. [](commit_id = 79e0e84df24c9435156952d54fbd48643d4e39a5, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>             static RefKind getRefKind(RefKind refKind) => refKind == RefKind.None ? RefKind.None : RefKind.Ref;

It looks like we should never get here with ```refKind == RefKind.RefReadOnlyParameter```. Consider adding an assert.

---
Refers to: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LoweredDynamicOperationFactory.cs:806 in 79e0e84. [](commit_id = 79e0e84df24c9435156952d54fbd48643d4e39a5, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

**Comment by jjonescz** [comment]
>         public static bool TryParse(string refKindString, int capacity, out RefKindVector result)

I don't think so. The only usage I found is in [anonymous type manager](https://github.com/dotnet/roslyn/blob/122b43d10a03a061e40f7348cfd461c332a5e238/src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs#L569) which parses names from the current module being built.

---
In reply to: [1638050803](https://github.com/dotnet/roslyn/pull/69034#issuecomment-1638050803) [](http://example.com/codeflow?ancestors=1638050803,1638050803)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Synthesized/RefKindVector.cs:109 in 79e0e84. [](commit_id = 79e0e84df24c9435156952d54fbd48643d4e39a5, deletion_comment = False)

**Comment by jjonescz** [comment]
>         public static bool TryParse(string refKindString, int capacity, out RefKindVector result)

> It enumerates `GetPreviousAnonymousDelegates`

Yes, but that's used during Edit & Continue, which I hope cannot span across different compiler versions.

But if we don't want to change `RefKindVector` encoding, I'm fine with that, I can close this PR and just remove the prototype comments - everything works fine - just delegates with `ref readonly` parameters are encoded as `<>f__AnonymousDelegate`s not `<>A{0000000}` delegates.

---
In reply to: [1638269099](https://github.com/dotnet/roslyn/pull/69034#issuecomment-1638269099) [](http://example.com/codeflow?ancestors=1638269099,1638050803)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Synthesized/RefKindVector.cs:109 in 79e0e84. [](commit_id = 79e0e84df24c9435156952d54fbd48643d4e39a5, deletion_comment = False)

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 2]
```diff
@@ -3085,4 +3085,100 @@ .maxstack  2
...
+                {
+                    dynamic d = 1;
+                    var c = new C();
+                    try
+                    {
+                        c.M(d);
+                    }
+                    catch (Microsoft.CSharp.RuntimeBinder.RuntimeBinderException)
+                    {
+                        System.Console.Write("exception");
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:3106]: nit: in the future, consider adding spaces at the end of such strings (that are meant to be concatenated into the output), I was looking for "exception2" in the source and couldn't find it, haha

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

**Pull request merged**